src/connection.rs

```diff
@@ -0,0 +1,180 @@
+                }
+            }
+        }))
+    }
+
+    fn start_serf_rx(&mut self) -> Result<JoinHandle<()>, String> {
+        let notifier = Arc::clone(&self.notifier);
+        let cancel = Arc::clone(&self.cancel);
+        let dispatch = Arc::clone(&self.dispatch);
+        let mut reader = BufReader::new(self.stream.try_clone().map_err(|x| x.to_string())?);
+        Ok(std::thread::spawn(move || {
+            loop {
+                let crate::protocol::ResponseHeader { seq, error } =
+                    match rmp_serde::from_read(&mut reader) {
+                        Ok(x) => x,
+                        Err(e) => {
+                            error!("Error while reading messagepack: {e}");
+                            cancel.store(true, std::sync::atomic::Ordering::Relaxed);
+                            notifier.notify_one();
+                            return;
+                        }
+                    };
+
+                let seq_handler = {
+                    let mut dispatch = dispatch.lock().unwrap();
+                    match dispatch.map.get(&seq) {
+                        Some(v) => {
+                            if v.streaming() {
+                                if !v.stream_acked() {
+                                    continue;
+                                }
+                                v.clone()
+                            } else {
+                                dispatch.map.remove(&seq).unwrap()
+                            }
+                        }
+                        None => {
+                            // response with no handler, ignore
+                            continue;
+                        }
+                    }
+                };
+
+                let res = if error.is_empty() {
+                    Ok(SeqRead(&mut reader))
+                } else {
+                    Err(error)
+                };
+
+                seq_handler.handle(res);
+            }
+        }))
+    }
+
+    #[tokio::main(flavor = "current_thread")]
+    async fn connection_watcher(
+        dispatch: Arc<Mutex<DispatchMap>>,
+        cancel: Arc<AtomicBool>,
+        thread_notifier: Arc<Notify>,
+        stream: TcpStream,
+        serf_tx_handle: JoinHandle<()>,
+        serf_rx_handle: JoinHandle<()>,
+    ) {
+        while !cancel.load(std::sync::atomic::Ordering::Relaxed) {
+            info!("Waiting until connection threads send notification");
+            thread_notifier.notified().await;
+        }
+
+        info!("Cleaning up SERF threads");
+        drop(stream);
+
+        info!("Cleaning up existing SERF requests by responding with an error");
+        let mut lock = dispatch.lock().unwrap();
+        let map = &mut lock.map;
+        for (_, value) in map.iter() {
+            value.handle(Err(
+                "Request was cancelled due to an error in the SERF connection".to_string(),
+            ));
+        }
+        map.clear();
+        drop(lock);
```

Comment:
I think this way the lock is dropped automatically as soon as the map is no longer used. Also, `drain` can be used to consume all the key/value pairs from the map.
```suggestion
        for (_, value) in dispatch.lock().unwrap().map.drain() {
            value.handle(Err(
                "Request was cancelled due to an error in the SERF connection".to_string(),
            ));
        }
```

Reply:
Yes, this will automatically drop the lock afterwards. I didn't know that `drain` existed so I had to split up the calls and used a manual `drop` to ensure the duration of the lock was as short as can be. Fixed.